### PR TITLE
Cleanup nll

### PR DESCRIPTION
**Pull request by ljedrz**
- improve allocations
- improve `format!` calls
- improve common patterns

### CODE REVIEW

File: src/librustc_mir/borrow_check/nll/constraint_generation.rs [hunk 1]
```diff
@@ -164,7 +165,9 @@ impl<'cg, 'cx, 'gcx, 'tcx> Visitor<'tcx> for ConstraintGeneration<'cg, 'cx, 'gcx
                 self.location_table.mid_index(location),
             ));
 
-            for successor_block in terminator.successors() {
+            let successor_blocks = terminator.successors();
+            all_facts.cfg_edge.reserve(successor_blocks.size_hint().0);
```
Comment (Mark-Simulacrum) [inline src/librustc_mir/borrow_check/nll/constraint_generation.rs]: Could presumably `count()` too since I think it should be cheap (`successors` returns `Chain<IntoIter, slice::Iter>`)

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 2]
```diff
@@ -254,7 +255,7 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         &self,
         borrow_location: Location,
     ) -> bool {
-        let mut visited_locations = Vec::new();
+        let mut visited_locations = FxHashSet::default();
```
Comment (Mark-Simulacrum) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: Hm, I'd rather make this change in a separate PR which would be benchmarked to make sure this is actually an optimization (it looks like it would be but I'm not sure).

**Review by estebank — APPROVED** [review]
LGTM, after addressing @Mark-Simulacrum's comments.

File: src/librustc_mir/borrow_check/nll/constraint_generation.rs (hunk 1 above)
Comment (ljedrz) [reply src/librustc_mir/borrow_check/nll/constraint_generation.rs]: Neat, I didn't know it was this cheap for `slice::Iter`.

Comment (Mark-Simulacrum) [reply src/librustc_mir/borrow_check/nll/constraint_generation.rs]: I'm fairly certain it is (count() = len())

**Comment by estebank** [comment]
```
[00:54:08] ---- [ui] ui/nll/polonius-smoke-test.rs stdout ----
[00:54:08] diff of stderr:
[00:54:08] 
[00:54:08] 28	LL |     y
[00:54:08] 29	   |     - borrow later used here
[00:54:08] 30	
[00:54:08] -	error[E0505]: cannot move out of `s` because it is borrowed
[00:54:08] -	  --> $DIR/polonius-smoke-test.rs:43:5
[00:54:08] -	   |
[00:54:08] -	LL |     let r = &mut *s;
[00:54:08] -	   |             ------- borrow of `*s` occurs here
[00:54:08] -	LL |     let tmp = foo(&r);
[00:54:08] -	LL |     s; //~ ERROR
[00:54:08] -	   |     ^ move out of `s` occurs here
[00:54:08] -	LL |     tmp;
[00:54:08] -	   |     --- borrow later used here
[00:54:08] -	
[00:54:08] -	error: aborting due to 4 previous errors
[00:54:08] +	error: aborting due to 3 previous errors
[00:54:08] 43	
[00:54:08] 44	Some errors occurred: E0503, E0505, E0597.
[00:54:08] 45	For more information about an error, try `rustc --explain E0503`.
```

**Comment by ljedrz** [comment]
Yes, but how? I only changed `size_hint` to `count`.

**Comment by Mark-Simulacrum** [comment]
Count will exhaust the iterator (by_ref doesn't clone it) regardless of whether it's implemented to be more efficient, I believe.

File: src/librustc_mir/borrow_check/nll/constraint_generation.rs [hunk 3]
```diff
@@ -164,7 +165,9 @@ impl<'cg, 'cx, 'gcx, 'tcx> Visitor<'tcx> for ConstraintGeneration<'cg, 'cx, 'gcx
                 self.location_table.mid_index(location),
             ));
 
-            for successor_block in terminator.successors() {
+            let mut successor_blocks = terminator.successors();
+            all_facts.cfg_edge.reserve(successor_blocks.by_ref().count());
+            for successor_block in successor_blocks {
```
Comment (estebank) [inline src/librustc_mir/borrow_check/nll/constraint_generation.rs]: > Yes, but how? I only changed size_hint to count (plus the related iterator stuff).

`.count()` consumes the iterator and you're trying to iterate over it after consuming it here.

Comment (ljedrz) [reply src/librustc_mir/borrow_check/nll/constraint_generation.rs]: Ow, I forgot that `by_ref` still advances the iterator; I'll revert that.
